File: android/src/main/java/com/applovin/reactnative/AppLovinMAXAdViewAdManager.java
package com.applovin.reactnative;

import android.text.TextUtils;

import com.applovin.mediation.MaxAdFormat;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.HashMap;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

@ReactModule(name = "AppLovinMAXAdViewAdManager")
class AppLovinMAXAdViewAdManager
        extends ReactContextBaseJavaModule
{
    private static final String ON_ADVIEW_AD_LOADED_EVENT         = "OnAdViewAdLoadedEvent";
    private static final String ON_ADVIEW_AD_LOAD_FAILED_EVENT    = "OnAdViewAdLoadFailedEvent";
    private static final String ON_ADVIEW_AD_CLICKED_EVENT        = "OnAdViewAdClickedEvent";
    private static final String ON_ADVIEW_AD_COLLAPSED_EVENT      = "OnAdViewAdCollapsedEvent";
    private static final String ON_ADVIEW_AD_EXPANDED_EVENT       = "OnAdViewAdExpandedEvent";
    private static final String ON_ADVIEW_AD_DISPLAY_FAILED_EVENT = "OnAdViewAdDisplayFailedEvent";
    private static final String ON_ADVIEW_AD_REVENUE_PAID         = "OnAdViewAdRevenuePaid";

    private final Map<Integer, AppLovinMAXAdViewAd> adViewAds = new HashMap<>( 2 );

    AppLovinMAXAdViewAdManager(final ReactApplicationContext context)
    {
        super( context );
    }

    @NonNull
    @Override
    public String getName()
    {
        return "AppLovinMAXAdViewAdManager";
    }

    public AppLovinMAXAdViewAd getAdViewAd(final int adViewAdId)
    {
        return adViewAds.get( adViewAdId );
    }

    @ReactMethod
    public void createAdView(final String adUnitId, final String adFormatStr, final Promise promise)
    {
        if ( TextUtils.isEmpty( adUnitId ) )
        {
            promise.reject( new IllegalStateException( "Attempting to create MaxAdView without Ad Unit ID" ) );
            return;
        }

        MaxAdFormat adFormat;

        if ( MaxAdFormat.BANNER.getLabel().equals( adFormatStr ) )
        {
            adFormat = AppLovinMAXModule.getDeviceSpecificBannerAdViewAdFormat( getReactApplicationContext() );
        }
        else if ( MaxAdFormat.MREC.getLabel().equals( adFormatStr ) )
        {
            adFormat = MaxAdFormat.MREC;
        }
        else
        {
            promise.reject( new IllegalStateException( "Attempting to set an invalid ad format of \"" + adFormatStr + "\" for " + adUnitId ) );
            return;
        }

        AppLovinMAXAdViewAd adViewAd = new AppLovinMAXAdViewAd( getReactApplicationContext(), this );
        int adViewAdId = adViewAd.createAdView( adUnitId, adFormat );
        adViewAds.put( adViewAdId, adViewAd );
        promise.resolve( adViewAdId );
    }

    @ReactMethod
    public void destroyAdView(final int adViewAdId, final Promise promise)
    {
        AppLovinMAXAdViewAd adViewAd = adViewAds.get( adViewAdId );
        adViewAd.destroyAdView();
        adViewAds.remove( adViewAdId );
        promise.resolve( null );
    }

    @ReactMethod
    public void loadAd(final int adViewAdId, final Promise promise)
    {
        AppLovinMAXAdViewAd adViewAd = adViewAds.get( adViewAdId );
        adViewAd.loadAd();
        promise.resolve( null );
    }

    @ReactMethod
    public void setPlacement(final int adViewAdId, final String placement, final Promise promise)
    {
        AppLovinMAXAdViewAd adViewAd = adViewAds.get( adViewAdId );
        adViewAd.setPlacement( placement );
        promise.resolve( null );
    }

    @ReactMethod
    public void setCustomData(final int adViewAdId, final String customData, final Promise promise)
    {
        AppLovinMAXAdViewAd adViewAd = adViewAds.get( adViewAdId );
        adViewAd.setCustomData( customData );
        promise.resolve( null );
    }

    @ReactMethod
    public void setAdaptiveBannerEnabled(final int adViewAdId, final boolean adaptiveBannerEnabled, final Promise promise)
    {
        AppLovinMAXAdViewAd adViewAd = adViewAds.get( adViewAdId );
        adViewAd.setAdaptiveBannerEnabled( adaptiveBannerEnabled );
        promise.resolve( null );
    }

    @ReactMethod
    public void setAutoRefresh(final int adViewAdId, final boolean autoRefresh, final Promise promise)
    {
        AppLovinMAXAdViewAd adViewAd = adViewAds.get( adViewAdId );
        adViewAd.setAutoRefresh( autoRefresh );
        promise.resolve( null );
    }

    @ReactMethod
    public void setExtraParameters(final int adViewAdId, final ReadableMap parameterMap, final Promise promise)
    {
        AppLovinMAXAdViewAd adViewAd = adViewAds.get( adViewAdId );
        adViewAd.setExtraParameters( parameterMap );
        promise.resolve( null );
    }

    @ReactMethod
    public void setLocalExtraParameters(final int adViewAdId, final ReadableMap parameterMap, final Promise promise)
    {
        AppLovinMAXAdViewAd adViewAd = adViewAds.get( adViewAdId );
        adViewAd.setLocalExtraParameters( parameterMap );
        promise.resolve( null );
    }

    public void sendLoadAdEvent(@Nullable final WritableMap params)
    {
        sendReactNativeEvent( ON_ADVIEW_AD_LOADED_EVENT, params );
    }

    public void sendFailToDisplayAEvent(@Nullable final WritableMap params)
    {
        sendReactNativeEvent( ON_ADVIEW_AD_DISPLAY_FAILED_EVENT, params );
    }

    public void sendFailToLoadAdEvent(@Nullable final WritableMap params)
    {
        sendReactNativeEvent( ON_ADVIEW_AD_LOAD_FAILED_EVENT, params );
    }

    public void sendClickAdEvent(@Nullable final WritableMap params)
    {
        sendReactNativeEvent( ON_ADVIEW_AD_CLICKED_EVENT, params );
    }

    public void sendCollapseAdEvent(@Nullable final WritableMap params)
    {
        sendReactNativeEvent( ON_ADVIEW_AD_COLLAPSED_EVENT, params );
    }

    public void sendExpandAdEvent(@Nullable final WritableMap params)
    {
        sendReactNativeEvent( ON_ADVIEW_AD_EXPANDED_EVENT, params );
    }

    public void sendPayRevenueEvent(@Nullable final WritableMap params)
    {
        sendReactNativeEvent( ON_ADVIEW_AD_REVENUE_PAID, params );
    }

    private void sendReactNativeEvent(final String name, @Nullable final WritableMap params)
    {
        getReactApplicationContext()
                .getJSModule( DeviceEventManagerModule.RCTDeviceEventEmitter.class )
                .emit( name, params );
    }

    @Override
    @Nullable
    public Map<String, Object> getConstants()
    {
        final Map<String, Object> constants = new HashMap<>();

        constants.put( "ON_ADVIEW_AD_LOADED_EVENT", ON_ADVIEW_AD_LOADED_EVENT );
        constants.put( "ON_ADVIEW_AD_LOAD_FAILED_EVENT", ON_ADVIEW_AD_LOAD_FAILED_EVENT );
        constants.put( "ON_ADVIEW_AD_CLICKED_EVENT", ON_ADVIEW_AD_CLICKED_EVENT );
        constants.put( "ON_ADVIEW_AD_COLLAPSED_EVENT", ON_ADVIEW_AD_COLLAPSED_EVENT );
        constants.put( "ON_ADVIEW_AD_EXPANDED_EVENT", ON_ADVIEW_AD_EXPANDED_EVENT );
        constants.put( "ON_ADVIEW_AD_DISPLAY_FAILED_EVENT", ON_ADVIEW_AD_DISPLAY_FAILED_EVENT );
        constants.put( "ON_ADVIEW_AD_REVENUE_PAID", ON_ADVIEW_AD_REVENUE_PAID );

        return constants;
    }
}

File: ios/AppLovinMAXAdViewAdManager.m
#import "AppLovinMAX.h"
#import "AppLovinMAXAdViewAdManager.h"
#import "AppLovinMAXAdViewAd.h"

static NSString *const ON_ADVIEW_AD_LOADED_EVENT = @"OnAdViewAdLoadedEvent";
static NSString *const ON_ADVIEW_AD_LOAD_FAILED_EVENT = @"OnAdViewAdLoadFailedEvent";
static NSString *const ON_ADVIEW_AD_CLICKED_EVENT = @"OnAdViewAdClickedEvent";
static NSString *const ON_ADVIEW_AD_COLLAPSED_EVENT = @"OnAdViewAdCollapsedEvent";
static NSString *const ON_ADVIEW_AD_EXPANDED_EVENT = @"OnAdViewAdExpandedEvent";
static NSString *const ON_ADVIEW_AD_DISPLAY_FAILED_EVENT = @"OnAdViewAdDisplayFailedEvent";
static NSString *const ON_ADVIEW_AD_REVENUE_PAID = @"OnAdViewAdRevenuePaid";

@interface AppLovinMAXAdViewAdManager()
@property (nonatomic, strong) NSMutableDictionary<NSNumber *, AppLovinMAXAdViewAd *> *adViewAds;
@end

@implementation AppLovinMAXAdViewAdManager

RCT_EXPORT_MODULE()

+ (BOOL)requiresMainQueueSetup
{
    return YES;
}

- (dispatch_queue_t)methodQueue
{
    return dispatch_get_main_queue();
}

- (instancetype)init
{
    self = [super init];
    if ( self )
    {
        self.adViewAds = [NSMutableDictionary dictionaryWithCapacity: 2];
    }
    return self;
}

- (AppLovinMAXAdViewAd *)adViewAd:(nonnull NSNumber *)adViewAdId
{
    return self.adViewAds[adViewAdId];
}

RCT_EXPORT_METHOD(createAdView:(NSString *)adUnitIdentifier :(NSString *)adFormatStr :(RCTPromiseResolveBlock)resolve :(RCTPromiseRejectBlock)reject)
{
    if ( ![adUnitIdentifier al_isValidString] )
    {
        reject(RCTErrorUnspecified, @"Attempting to attach MAAdView without Ad Unit ID", nil);
        return;
    }

    MAAdFormat *adFormat;
    
    if ( [MAAdFormat.banner.label isEqualToString: adFormatStr] )
    {
        adFormat = DEVICE_SPECIFIC_ADVIEW_AD_FORMAT;
    }
    else if ( [MAAdFormat.mrec.label isEqualToString: adFormatStr] )
    {
        adFormat = MAAdFormat.mrec;
    }
    else
    {
        NSString *error = [NSString stringWithFormat: @"Attempting to set an invalid ad format of %@ for %@!", adFormatStr, adUnitIdentifier];
        reject(RCTErrorUnspecified, error, nil);
        return;
    }
    
    AppLovinMAXAdViewAd *adViewAd = [[AppLovinMAXAdViewAd alloc] initWithManager: self];
    NSNumber *adViewAdId = [adViewAd createAdView: adUnitIdentifier adFormat: adFormat];
    self.adViewAds[adViewAdId] = adViewAd;
    resolve(adViewAdId);
}

RCT_EXPORT_METHOD(destroyAdView:(nonnull NSNumber *)adViewAdId :(RCTPromiseResolveBlock)resolve :(RCTPromiseRejectBlock)reject)
{
    AppLovinMAXAdViewAd *adViewAd = self.adViewAds[adViewAdId];
    [adViewAd destroyAdView];
    [self.adViewAds removeObjectForKey: adViewAdId];
    resolve(nil);
}

RCT_EXPORT_METHOD(loadAd:(nonnull NSNumber *)adViewAdId :(RCTPromiseResolveBlock)resolve :(RCTPromiseRejectBlock)reject)
{
    AppLovinMAXAdViewAd *adViewAd = self.adViewAds[adViewAdId];
    [adViewAd loadAd];
    resolve(nil);
}

RCT_EXPORT_METHOD(setPlacement:(nonnull NSNumber *)adViewAdId :(nullable NSString *)placement :(RCTPromiseResolveBlock)resolve :(RCTPromiseRejectBlock)reject)
{
    AppLovinMAXAdViewAd *adViewAd = self.adViewAds[adViewAdId];
    adViewAd.placement = placement;
    resolve(nil);
}

RCT_EXPORT_METHOD(setCustomData:(nonnull NSNumber *)adViewAdId :(nullable NSString *)customData :(RCTPromiseResolveBlock)resolve :(RCTPromiseRejectBlock)reject)
{
    AppLovinMAXAdViewAd *adViewAd = self.adViewAds[adViewAdId];
    adViewAd.customData = customData;
    resolve(nil);
}

RCT_EXPORT_METHOD(setAdaptiveBannerEnabled:(nonnull NSNumber *)adViewAdId :(BOOL)adaptiveBannerEnabled :(RCTPromiseResolveBlock)resolve :(RCTPromiseRejectBlock)reject)
{
    AppLovinMAXAdViewAd *adViewAd = self.adViewAds[adViewAdId];
    adViewAd.adaptiveBannerEnabled = adaptiveBannerEnabled;
    resolve(nil);
}

RCT_EXPORT_METHOD(setAutoRefresh:(nonnull NSNumber *)adViewAdId :(BOOL)autoRefresh :(RCTPromiseResolveBlock)resolve :(RCTPromiseRejectBlock)reject)
{
    AppLovinMAXAdViewAd *adViewAd = self.adViewAds[adViewAdId];
    adViewAd.autoRefresh = autoRefresh;
    resolve(nil);
}

RCT_EXPORT_METHOD(setExtraParameters:(nonnull NSNumber *)adViewAdId :(NSDictionary<NSString *, id> *)parameterDict :(RCTPromiseResolveBlock)resolve :(RCTPromiseRejectBlock)reject)
{
    AppLovinMAXAdViewAd *adViewAd = self.adViewAds[adViewAdId];
    adViewAd.extraParameters = parameterDict;
    resolve(nil);
}

RCT_EXPORT_METHOD(setLocalExtraParameters:(nonnull NSNumber *)adViewAdId :(NSDictionary<NSString *, id> *)parameterDict :(RCTPromiseResolveBlock)resolve :(RCTPromiseRejectBlock)reject)
{
    AppLovinMAXAdViewAd *adViewAd = self.adViewAds[adViewAdId];
    adViewAd.localExtraParameters = parameterDict;
    resolve(nil);
}

- (void)sendLoadAdEvent:(NSDictionary<NSString *, id> *)body
{
    [self sendEventWithName: ON_ADVIEW_AD_LOADED_EVENT body: body];
}

- (void)sendFailToDisplayAEvent:(NSDictionary<NSString *, id> *)body
{
    [self sendEventWithName: ON_ADVIEW_AD_DISPLAY_FAILED_EVENT body: body];
}

- (void)sendFailToLoadAdEvent:(NSDictionary<NSString *, id> *)body
{
    [self sendEventWithName: ON_ADVIEW_AD_LOAD_FAILED_EVENT body: body];
}

- (void)sendClickAdEvent:(NSDictionary<NSString *, id> *)body
{
    [self sendEventWithName: ON_ADVIEW_AD_CLICKED_EVENT body: body];
}

- (void)sendCollapseAdEvent:(NSDictionary<NSString *, id> *)body
{
    [self sendEventWithName: ON_ADVIEW_AD_COLLAPSED_EVENT body: body];
}

- (void)sendExpandAdEvent:(NSDictionary<NSString *, id> *)body
{
    [self sendEventWithName: ON_ADVIEW_AD_EXPANDED_EVENT body: body];
}

- (void)sendPayRevenueEvent:(NSDictionary<NSString *, id> *)body
{
    [self sendEventWithName: ON_ADVIEW_AD_REVENUE_PAID body: body];
}

- (NSArray<NSString *> *)supportedEvents
{
    return @[ON_ADVIEW_AD_LOADED_EVENT,
             ON_ADVIEW_AD_LOAD_FAILED_EVENT,
             ON_ADVIEW_AD_CLICKED_EVENT,
             ON_ADVIEW_AD_COLLAPSED_EVENT,
             ON_ADVIEW_AD_EXPANDED_EVENT,
             ON_ADVIEW_AD_DISPLAY_FAILED_EVENT,
             ON_ADVIEW_AD_REVENUE_PAID];
}

- (NSDictionary *)constantsToExport
{
    return @{@"ON_ADVIEW_AD_LOADED_EVENT" : ON_ADVIEW_AD_LOADED_EVENT,
             @"ON_ADVIEW_AD_LOAD_FAILED_EVENT" : ON_ADVIEW_AD_LOAD_FAILED_EVENT,
             @"ON_ADVIEW_AD_CLICKED_EVENT" : ON_ADVIEW_AD_CLICKED_EVENT,
             @"ON_ADVIEW_AD_COLLAPSED_EVENT" : ON_ADVIEW_AD_COLLAPSED_EVENT,
             @"ON_ADVIEW_AD_EXPANDED_EVENT" : ON_ADVIEW_AD_EXPANDED_EVENT,
             @"ON_ADVIEW_AD_DISPLAY_FAILED_EVENT" : ON_ADVIEW_AD_DISPLAY_FAILED_EVENT,
             @"ON_ADVIEW_AD_REVENUE_PAID" : ON_ADVIEW_AD_REVENUE_PAID};
}

@end
